enhancements/node-healthcheck-operator.md

```diff
@@ -0,0 +1,435 @@
+1. health criteria: list of criterias and selector
+2. template: a template of an [external custom resource (remediation CR)](link todo)
+ to create in response to a match.
+
+### Unhealthy criteria
+A node target is unhealthy when the node meets the unhealthy node conditions criteria defined.
+If any of those criterias are met for longer than given timeouts, the controller
+creates a remediation CR according to the template in the NHC CRD.
+Timeouts will be defined as part of the NHC CRD .
+
+### Implementation Details
+
+#### NodeHealthCheck CRD
+- Enable watching a node based on a label selector. 
+- Enable defining an unhealthy node criteria (based on a list of node conditions).
+- Enable setting a threshold of unhealthy nodes. If the current number is at or above this threshold no further action will take place. This can be expressed as an int or as a percentage of the total targets in the pool.
+
+E.g:
+- Create a remedy CR when node is `ready=false` or `ready=Unknown` condition for more than 5m.
+- I want to avoid triggering remediation if 40% or more of the targets of this pool are unhealthy at the same time.
+
+
+```yaml
+apiVersion: remediation.medik8s.io/v1alpha1
+kind: NodeHealthCheck
+metadata:
+  name: example
+  # this CRD is cluster-scope, namespace is not needed  
+spec:
+  maxUnhealthy: "49%"
+  selector:
+    matchExpressions:
+      key: node-role.kubernetes.io/worker
+      operator: exists
+  unhealthyConditions:
+    - type: "Ready"
+      status:  "Unknown"
+      duration: "300s"
+   - type: "Ready"
+     status:  "False"
+     duration: "300s"
+  remediationTemplate:
+    kind: PoisonPillRemediationTemplate
+    apiVersion: poison-pill.medik8s.io/v1alphaX
+    name: M3_REMEDIATION_GROUP
+status:
+  healthyNodes: 5
+  obeservedNodes: 5
+```
+
+#### NodeHealthCheck controller
+Watch:
+- Watch NodeHealthCheck resources
+- Watch nodes with an event handler e.g controller runtime `EnqueueRequestsFromMapFunc` which returns NodeHealthCheck resources.
+
+Reconcile:
+- Periodically enqueue a reconcile request
+- Fetch all nodes targets. E.g:
+
+
+- Calculate the number of unhealthy targets.
+- Compare current number against `maxUnhealthy` threshold and temporary short
+ circuits if the threshold is met.
+- Create remedy objects for unhealthy targets according to the template in NHC object
+
+### Notes/Constraints
+
+The functionality of this controller have a lot in common with the Machine Health
+Check controller except that its creating objects and not
+performing any actions per se. When deploying this controller make
+sure 1. you have a controller implementing the API, e.g [poison pill]
+2. the machine health check controller is not running, or in case of a mixed
+ the node health checks selector avoid targeting nodes with machines.
+
+
+### Risks and Mitigations
+
+#### Remediation Interaction with ClusterAutoScalar
+
+Misconfiguration of nodes provisioning can cause nodes to fail to get operational.
```

Comment:
If the node doesn't become operational, how will the NHC remediate this? I don't really understand how the NHC would have this kind of issue at all, it doesn't know which nodes are meant to be coming up, which is where this issue occurs in MHC

Reply:
NHC's problem is being able to detect such a condition, remediation is delegated to external operators.
As discussed below, today NHC doesn't detect this scenario and is one reason why it doesn't replace MHC.

Even if NHC did detect the condition, the primary goal is to ensure that protected workloads are not running there (so that they can be started elsewhere). Which, even if the NHC does nothing, is true by default for a node that has never contacted the scheduler.

While restoring compute capacity in the presence of non-transient errors is not possible without MAPI[1], this doesn't detract from being able to automatically and safely move the workloads elsewhere.

[1] And sometimes even *with* MAPI.  Reprovisioning a physical node with a bad disk or DIMM is going to give you back the same broken machine.

Reply:
I understand that NHC doesn't detect this condition, but the topic of this paragraph seems to imply it is borrowing the back off mechanism implemented in MHC for this very reason.

Given NHC only knows about Nodes and doesn't know about their owners, i don't see how it can track or relate nodes in a way that a backoff could be implemented, unless this was a global backoff for the entire NHC?

How often do we predict that this kind of backoff would be required?

Reply:
I think it's more trying to say why we *can't* borrow the same exact mechanism from MHC.

Since NHC is primarily useful on clusters without MAPI, the Node would have the same name when it comes back though - tracking how many times the same node was remediated is possible in those circumstances, for anywhere else it could be handled globally (rg. I remediated N nodes in M minutes, I should stop)

Reply:
That's good context, please make sure that's reflected into the enhancement in this section